### DISCUSSION TITLE
Port RAIntegration from Win32 API to QT

### DISCUSSION THREAD

**Discussion by Jamiras** [RetroAchievements/RAIntegration#763] [Ideas]
Replace the Windows-specific functionality in the DLL with something that can be used on other platforms.

_Extracted from #64_

> This is a long-term goal. I recognize that this requires a good amount of effort, but I want to highlight this as a goal in a hope to prevent the current development from moving toward a more intensive use of Windows-only solutions.
>
> We know that some libretro guys are interested on this QT port, and maybe they can help here at some point.
>
> Being able to develop achievements on Linux is like a dream for me and I'm sure that it would also bring many knowledgeable people to the scene.
>
> IMHO making the cheevo development toolkit be cross-platform is a huge contribution to its longevity.

Note that QT is just a suggestion, and there may be a better alternative.

**Comment by Blakeinstein** [comment]
From what I understand from this post, is that the only the development is restricted to windows? What If I wanted to integrate cheevos in my own emulator, is it possible and/or allowed? And if so, where do I start?

I believe it is possible cross-platform since RetroArch is cross-platform, but I would like support for the same in an IOS-only emulator, Delta, purely because of its visual appeal and platform optimizations.

**Reply by Blakeinstein** [reply]
Oh and sorry for reviving an old discussion.

**Reply by Jamiras** [reply]
RAIntegration is the DLL that we use to _create_ achievements. It is currently only available on Windows because of it's dependency on the Win32 APIs. This discussion is a feature request to refactor the code so that the Win32 API dependencies can be replaced with something else - QT was the original suggestion.

If you only need _runtime_ support, look at [rcheevos](https://github.com/RetroAchievements/rcheevos). That's what RetroArch uses. It's completely platform-independent C code that you can compile directly into your project. You'll need to provide a networking layer (that can make HTTP POST calls) and all the UI widgets. There's a [wiki](https://github.com/RetroAchievements/rcheevos/wiki) documenting the API functions, but no real guide for doing the implementation. The only things we ask are that you provide a distinguishing User-Agent (i.e. MyEmulator/1.0) and that you honor the hardcore restrictions (no loading save states, no rewind, no viewing memory, etc) if you submit any hardcore unlocks.

If you're looking just to display information available from the website (without actually having active achievements), we also have several [public JSON APIs](https://github.com/RetroAchievements/RAWeb/tree/master/public/API). Note that these don't support user login. Instead, they require a WebAPI key (available from the user's settings page).
